cpp/src/spatial/haversine.cu

```diff
@@ -0,0 +1,153 @@
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("haversine_distance does not support non-floating-point types.");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(cudf::column_view const& a_lon,
+               cudf::column_view const& a_lat,
+               cudf::column_view const& b_lon,
+               cudf::column_view const& b_lat,
+               double radius,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+    {
+        auto size = a_lon.size();
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto result = cudf::make_fixed_width_column(cudf::data_type{ tid },
+                                                    size,
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+
+        if (size == 0)
+        {
+            return result;
+        }
+
+        auto input_tuple = thrust::make_tuple(thrust::make_constant_iterator<T>(radius),
+                                              a_lon.begin<T>(),
+                                              a_lat.begin<T>(),
+                                              b_lon.begin<T>(),
+                                              b_lat.begin<T>());
+
+        auto input_iter = thrust::make_zip_iterator(input_tuple);
+
+        thrust::transform(rmm::exec_policy(stream)->on(stream),
+                          input_iter,
+                          input_iter + size,
+                          result->mutable_view().begin<T>(),
+                          [] __device__ (auto inputs) {
+                              return calculate_haversine_distance(thrust::get<0>(inputs),
+                                                                  thrust::get<1>(inputs),
+                                                                  thrust::get<2>(inputs),
+                                                                  thrust::get<3>(inputs),
+                                                                  thrust::get<4>(inputs));
+                          });
+
+        return result;
+    }
+};
+
+} // anonymous namespace
+
+namespace cuspatial {
+
+namespace detail {
+
+std::unique_ptr<cudf::column> haversine_distance(cudf::column_view const& a_lon,
+                                                 cudf::column_view const& a_lat,
+                                                 cudf::column_view const& b_lon,
+                                                 cudf::column_view const& b_lat,
+                                                 double radius,
+                                                 rmm::mr::device_memory_resource* mr,
+                                                 cudaStream_t stream)
```

Comment:
In libcudf we are in the process of correcting all the mis-ordered APIs. Should be stream, then mr.

```suggestion
                                                 cudaStream_t stream,
                                                 rmm::mr::device_memory_resource* mr)
```

Reply:
Ah, I recalled a conversation about this a while ago and thought we settled on `mr` _then_ `stream`, so I've been doing it that way. :facepalm:

Reply:
There's a PR in progress from @trevorsm7 .